app/src/main/java/com/hover/stax/login/LoginViewModel.kt

```diff
@@ -30,189 +33,192 @@ private const val BOUNTY_EMAIL_KEY = "email_for_bounties"
+            progress.value = 66
+            viewModelScope.launch(Dispatchers.IO) {
+                try {
+                    val result = LoginNetworking(application).uploadUserToStax(email, optedIn.value!!, token)
+                    Timber.e("Uploading user to stax came back: ${result.code}")
+
+                    if (result.code in 200..299) onSuccess(
+                        JSONObject(result.body!!.string()),
+                        application.getString(
+                            R.string.uploaded_to_hover,
+                            application.getString(R.string.upload_user)
+                        )
+                    )
+                    else onError(application.getString(R.string.upload_user_error))
+                } catch (e: IOException) {
+                    onError(application.getString(R.string.upload_user_error))
+                }
+            }
+        }
+    }
+
+    fun uploadLastUser() {
+        val account = GoogleSignIn.getLastSignedInAccount(application)
+        if (account != null) uploadUserToStax(email.value, account.idToken)
+        else Timber.e("No account found")
+    }
+
+    fun saveReferee(refereeCode: String, name: String, phone: String) {
+        viewModelScope.launch(Dispatchers.IO) {
+            if (!email.value.isNullOrEmpty()) {
+                val account = GoogleSignIn.getLastSignedInAccount(application)
+                Timber.i("save in referee method, now trying")
+                try {
+                    val result = LoginNetworking(application).uploadReferee(
+                        email.value!!,
+                        refereeCode,
+                        name,
+                        phone,
+                        account?.idToken
+                    )
+                    if (result.code in 200..299) onSuccess(JSONObject(result.body!!.string()), name)
+                    else onError(application.getString(R.string.upload_referee_error))
+                } catch (e: IOException) {
+                    onError(application.getString(R.string.upload_referee_error))
+                }
+            }
+        }
+    }
+
+    private fun onSuccess(json: JSONObject, name: String) {
+        Timber.e(json.toString())
+
+        refereeCode.value?.let {
+            val data = JSONObject()
+            try {
+                data.put("referee ID", it)
+                data.put("username", name)
+                AnalyticsUtil.logAnalyticsEvent(application.getString(R.string.upload_referee_success_event), data, application)
+            } catch (e: Exception) {
+                Timber.e(e.localizedMessage)
+            }
+        }
+
+        progress.postValue(100)
+        saveResponseData(json)
+    }
+
+    private fun setUser(firebaseUser: FirebaseUser, idToken: String) {
+        Timber.e("setting user: %s", firebaseUser.email)
+        user.postValue(firebaseUser)
+        setEmail(firebaseUser.email)
+
+        uploadUserToStax(firebaseUser.email, idToken)
```

Comment:
I removed the firebase `token` fetch here and called the `uploadUserToStax` method directly.

Reply:
Got it, thanks!

We probably might want to refractor the login view model a little bit in another story.
For example: SignInFirebaseAsync is doing like >3 things other than just signing in. 

Just an observation while reading.

This particular work is great, thanks.